### PR TITLE
Feature request: singleflight to return refcount vs shared bool

### PR DESCRIPTION
**Pull request by gleonid**
### Background:
We are using singleflight to run expensive procedure that creates a temporary resource (temp file)
Our challenge - we need to determine when it is safe to cleanup this temp resource
Currently `Group.Do/DoChan` calls only return a `shared bool` indicator, but they do not surface to caller how many actual caller will be using this result.

### Proposal to consider:
Currently implementation maintains "dups int" property for each `call`, semantically it is almost exactly a "reference counter" except in case of single caller it holds value 0 instead of 1.
I propose, to slightly change signature of `Group.Do`/`DoChan`: in addition to "shared bool" also return the actual reference counter (pointer to `call.dups` ?)

This PR provides an implementation for above
In short it does 2 things:
- defines a `refShared` type and changes `Group.Do` signature to use in lace of `shared bool`
- "refactor Group.Do implement via Group.Dochan call" (commit 4eee079)
   This slightly reduces duplicated code and will prevent a race condition on computation of `shared` flag. This race condition is introduced by previous item, as calculation of `shared bool` depends on value `call.dups` and that value could get changed by caller now.

### Note
It might be easier to review this PR on per commit basis

### CODE REVIEW

**Comment by gleonid** [comment]
@googlebot I signed it!

**Comment by gleonid** [comment]
declining this PR in favor of https://github.com/golang/sync/pull/9
